File: 06/Part1.java
import java.util.*;

public class Part1 {
  // Method to calculate the range
  private static Integer calculateRange(int T, int limit) {
    int start = -1, end = -1;
    // Loop from 0 to T
    for (int H = 0; H <= T; H += 1) {
        // Check if (T - H) * H is greater than limit
        if ((T - H) * H > limit) {
            // If start is -1, set it to H
            if (start == -1) {
                start = H;
            }
            // Update end to H
            end = H;
        }
    }
    // Return the range
    return end - start + 1;
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    List<Integer> timeList = new ArrayList<>();
    List<Integer> distanceList = new ArrayList<>();
    List<Integer> rangeList = new ArrayList<>();

    // Read lines from the input
    while (scanner.hasNextLine()) {
      String line = scanner.nextLine();
      String[] parts = line.split("\\s+");

      // If the line starts with "Time:", add the numbers to timeList
      if (line.startsWith("Time:")) {
        for (int i = 1; i < parts.length; i++) {
          timeList.add(Integer.parseInt(parts[i]));
        }
      }
      // If the line starts with "Distance:", add the numbers to distanceList
      else if (line.startsWith("Distance:")) {
        for (int i = 1; i < parts.length; i++) {
          distanceList.add(Integer.parseInt(parts[i]));
        }
      }
    }

    // Close the scanner
    scanner.close();

    // Calculate the range for each pair of time and distance, and add it to rangeList
    for (int i = 0; i < timeList.size(); i++) {
      rangeList.add(calculateRange(timeList.get(i), distanceList.get(i)));
    }

    // Print out the lists
    System.out.println("Time: " + timeList);
    System.out.println("Distance: " + distanceList);
    System.err.println("Range: " + rangeList);

    // Calculate the product of all elements in rangeList
    int product = 1;
    for (int num : rangeList) {
        product *= num;
    }
    // Print out the product
    System.out.println("Product of Range: " + product);
  }
}

File: 09/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	// Create a new scanner that reads from standard input
	scanner := bufio.NewScanner(os.Stdin)

	// Create a slice to hold the lines
	var lines [][]int64
	var total int64

	// Loop over all lines in standard input
	for scanner.Scan() {
		// Split the line into words
		words := strings.Fields(scanner.Text())
		// Add the line to the lines
		lines = append(lines, parseRow(words))
	}

	// Check for errors in the scanner
	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, "Error reading from stdin:", err)
	}

	// For each line, find the difference between all the numbers
	for _, line := range lines {
		// fmt.Println(line)

		differences := lineDifferences(line)
		// fmt.Println(differences)

		total += line[len(line)-1] + differences[len(differences)-1]

		// While not all numbers in the line are the same
		for !allSame(differences) {
			differences = lineDifferences(differences)
			total += differences[len(differences)-1]
		}
	}

	// Print the last elements
	fmt.Println(total)

}

// Function to check if all numbers in a slice are the same
func allSame(nums []int64) bool {
	for i := 1; i < len(nums); i++ {
		if nums[i] != nums[0] {
			return false
		}
	}
	return true
}

func parseRow(words []string) []int64 {
	// Convert each word to an integer and add it to the line
	var line []int64
	for _, word := range words {
		num, err := strconv.ParseInt(string(word), 10, 64)
		if err != nil {
			fmt.Fprintln(os.Stderr, "Error converting to integer:", err)
			continue
		}
		line = append(line, num) // Convert num to int64 before appending
	}
	return line
}

func lineDifferences(line []int64) []int64 {
	// Calculate the differences
	var nextLine []int64
	for i := 0; i < len(line)-1; i++ {
		nextLine = append(nextLine, line[i+1]-line[i])
	}
	return nextLine
}
